### PR TITLE
[Kernel] Add Split-KV Attention Kernel to the triton_attn Backend

### PR DESCRIPTION
**Pull request by jvlunteren**
In this PR we target performance improvements for the V1 `triton_attn` backend by adding a split-kv attention kernel (referred to as 3D kernel in the following) that accelerates the decoding by also parallelizing in the sequence dimension, which offers a clear advantage over the current upstream version (referred to as 2D kernel in the following), particularly in scenarios with small batch sizes and long sequences. Experiments have shown that the 3D kernel also outperforms the 2D kernel at large batch sizes.

## Performance
The following results were obtained for `meta-llama/Llama-3.1-8B-Instruct` on an NVIDIA H100 GPU, by running

<pre> $ VLLM_ATTENTION_BACKEND=TRITON_ATTN_VLLM_V1 VLLM_USE_V1=1 python benchmarks/benchmark_latency.py \
    --model meta-llama/Llama-3.1-8B-Instruct \
    --input-len 500 --output-len &ltoutput-length&gt \
    --batch-size &ltbatch-size&gt </pre>

for
1. the V1 `FlashAttention` backend (`VLLM_ATTENTION_BACKEND=TRITON_ATTN_VLLM_V1` is not used in the above command),
2.  the current triton 2D attention kernel, and
3. the new triton 3D attention kernel (this PR).

Results for batch sizes 1 and 128 are shown in the following two graphs. The output length (in tokens) was varied in these experiments across the following values: 10, 100, 200, 400, 800, 1600, 3200, 6400, and 12800 (the latter was only evaluated for batch size 1). The number of warmup iterations and measurement iterations were left at the default values of 10 and 30 respectively.

batch-size_1

batch-size_128

Note that in the upper figure (batch-size 1), the green and red lines overlap, indicating that the performance of the V1 `FlashAttention`  kernel and the 3D kernel is nearly identical.

This PR improves the performance of the V1 `triton_attn` backend by up to a factor of 3 for batch-size 1 as compared to the current 2D-kernel based implementation, while the 3D-kernel based implementation stays also closer in performance to the V1 `FlashAttention` kernel, even with longer batch sizes.

Further results for `benchmark_serving.py` (which only includes sequence lengths less than 2K tokens):

V1 `FlashAttention`:
<pre> $ python benchmarks/benchmark_serving.py \
    --model meta-llama/Llama-3.1-8B-Instruct \
    --dataset-name sharegpt \
    --dataset-path ShareGPT_V3_unfiltered_cleaned_split.json
============ Serving Benchmark Result ============
Successful requests:                     1000
Benchmark duration (s):                  20.94
Total input tokens:                      215196
Total generated tokens:                  198001
Request throughput (req/s):              47.75
Output token throughput (tok/s):         9453.78
Total Token throughput (tok/s):          19728.54
---------------Time to First Token----------------
Mean TTFT (ms):                          3386.15
Median TTFT (ms):                        3272.63
P99 TTFT (ms):                           6249.82
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          82.82
Median TPOT (ms):                        47.45
P99 TPOT (ms):                           215.18
---------------Inter-token Latency----------------
Mean ITL (ms):                           37.82
Median ITL (ms):                         26.81
P99 ITL (ms):                            218.57 
==================================================</pre>

triton 2D kernel:
<pre> $ python benchmarks/benchmark_serving.py \
    --model meta-llama/Llama-3.1-8B-Instruct \
    --dataset-name sharegpt \
    --dataset-path ShareGPT_V3_unfiltered_cleaned_split.json
============ Serving Benchmark Result ============
Successful requests:                     1000
Benchmark duration (s):                  25.11
Total input tokens:                      215196
Total generated tokens:                  197107
Request throughput (req/s):              39.83
Output token throughput (tok/s):         7850.21
Total Token throughput (tok/s):          16420.85
---------------Time to First Token----------------
Mean TTFT (ms):                          3521.11
Median TTFT (ms):                        3456.28
P99 TTFT (ms):                           6706.20
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          91.82
Median TPOT (ms):                        53.04
P99 TPOT (ms):                           238.21
---------------Inter-token Latency----------------
Mean ITL (ms):                           43.69
Median ITL (ms):                         31.70
P99 ITL (ms):                            242.32
================================================== </pre>

triton 3D kernel (this PR):
<pre> $ python benchmarks/benchmark_serving.py \
    --model meta-llama/Llama-3.1-8B-Instruct \
    --dataset-name sharegpt \
    --dataset-path ShareGPT_V3_unfiltered_cleaned_split.json
============ Serving Benchmark Result ============
Successful requests:                     1000
Benchmark duration (s):                  23.44
Total input tokens:                      215196
Total generated tokens:                  197248
Request throughput (req/s):              42.67
Output token throughput (tok/s):         8416.44
Total Token throughput (tok/s):          17598.70
---------------Time to First Token----------------
Mean TTFT (ms):                          4378.49
Median TTFT (ms):                        4312.11
P99 TTFT (ms):                           7586.13
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          92.06
Median TPOT (ms):                        51.45
P99 TPOT (ms):                           266.57
---------------Inter-token Latency----------------
Mean ITL (ms):                           41.43
Median ITL (ms):                         28.86
P99 ITL (ms):                            245.03
================================================== </pre>

## Correctness

V1 `FlashAttention`:
<pre>VLLM_USE_V1=1 lm_eval --model vllm --model_args pretrained=meta-llama/Llama-3.1-8B-Instruct --tasks gsm8k --num_fewshot 5 --batch_size auto --limit 500

|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value|   |Stderr|
|-----|------:|----------------|-----:|-----------|---|----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.788|±  |0.0183|
|     |       |strict-match    |     5|exact_match|↑  |0.766|±  |0.0190| </pre>

This PR:
<pre>VLLM_USE_V1=1 VLLM_ATTENTION_BACKEND=TRITON_ATTN_VLLM_V1 lm_eval --model vllm --model_args pretrained=meta-llama/Llama-3.1-8B-Instruct --tasks gsm8k --num_fewshot 5 --batch_size auto --limit 500
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value|   |Stderr|
|-----|------:|----------------|-----:|-----------|---|----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.798|±  |0.0180|
|     |       |strict-match    |     5|exact_match|↑  |0.778|±  |0.0186| </pre>

## How is this performance improvement achieved?

The original triton attention kernel uses a 2D grid to parallelize across batches and heads. To scale this further, we introduced a 3D grid for parallelizing over the sequence length. This is achieved by splitting the attention computation into multiple segments in the sequence dimension, and using online softmax to compute partial results in each segment. A reduction kernel is then used to combine the partial results, enabling correct normalization across the segment boundaries.

### CODE REVIEW

**Comment by jvlunteren** [comment]
I don't understand why the pre-commit check fails: it complains about line 413 in `chunked_prefill_paged_decode.py` being too long, but that is not the case. It seems like the pre-commit check uses an older version that was cached (I had the same problem with a previous PR, which I closed and now resubmitted, hoping to fix this issue).

**Comment by jvlunteren** [comment]
Above issues has now been resolved.

**Comment by jvlunteren** [comment]
Above issue with the pre-commit check has now been resolved.

**Review by LucasWilkinson — APPROVED** [review]

**Comment by maleksan85** [comment]
is it showing the same perf gain on MI300?
